source/common/common/utility.h

```diff
@@ -632,4 +633,64 @@ template <class Value> struct TrieLookupTable {
   TrieEntry<Value> root_;
 };
 
+// Helper class for allocating classes with variable-sized inlined storage. See
+// InlineString below for usage.
+class InlineStorage : public NonCopyable {
+public:
+  // Custom delete operator to keep C++14 from using the global operator delete(void*, size_t),
+  // which would result in the compiler error:
+  // "exception cleanup for this placement new selects non-placement operator delete"
+  static void operator delete(void* address) { ::operator delete(address); }
+
+protected:
+  static void* operator new(size_t object_size, size_t data_size) {
+    return ::operator new(object_size + data_size);
+  }
+};
+
+class InlineString;
+using InlineStringPtr = std::unique_ptr<InlineString>;
+
+// Represents immutable string data, keeping the storage inline with the
+// object. These cannot be copied or held by value; they must be created
+// as unique pointers.
+class InlineString : public InlineStorage {
+public:
+  /**
+   * @param str the string_view for which to create an InlineString
+   * @return a unique_ptr to the InlineString containing the bytes of str.
+   */
+  static InlineStringPtr create(absl::string_view str) {
+    return InlineStringPtr(new (str.size()) InlineString(str.data(), str.size()));
+  }
+
+  /**
+   * @return a std::string copy of the InlineString.
+   */
+  std::string toString() const { return std::string(data_, size_); }
+
+  /**
+   * @return a string_view into the InlineString.
+   */
+  absl::string_view toStringView() const { return absl::string_view(data_, size_); }
+
+  /**
+   * @return the number of bytes in the string
+   */
+  size_t size() const { return size_; }
+
+  /**
+   * @return a pointer to the first byte of the string.
+   */
+  const char* data() const { return data_; }
+
+private:
+  // Constructor is declared private so that no one constructs one without the
+  // proper size allocation. to accommodate the variable-size buffer.
+  InlineString(const char* str, size_t size) : size_(size) { memcpy(data_, str, size); }
+
+  size_t size_;
+  char data_[];
```

Comment:
Note that depends on what data members are placed before data_, data_ may not point to the address after *this, instead it can point to a address in the padding portion of *this. It may not be what we want.

<pre>
Address (low to high):    ----------------------------------------------->>
                          |                         |                     |
                          |        sizeof(T)        |    buffer_size      |
                                                |
                                             data_
</pre>

Reply:
Understood. I'm not 100% sure if's worth the verbosity, but I added this comment above the declaration for InlineStorage:

```
// Perf note: The alignment will be correct and safe without further
// consideration as long as there are no casts. But for micro-optimization,
// consider this case:
//   struct MyStruct : public InlineStorage { uint64_t a_; uint16_t b_; uint8_t data_[]; };
// When compiled with a typical compiler on a 64-bit machine:
//   sizeof(MyStruct) == 16, because the compiler will round up from 10 for uint64_t alignment.
// So:
//   calling new (6) MyStruct() causes an allocation of 16+6=22, rounded up to 24 bytes.
// But data_ doesn't need 8-byte alignment, so it will wind up adjacent to the uint16_t.
//   ((char*) my_struct.data) - ((char*) &my_struct) == 10
// If we had instead declared data_[6], then the whole allocation would have fit in 16 bytes.
// Instead:
//   - the starting address of data will not be 8-byte aligned. This is not required
//     by the C++ standard for a uint8_t, but may be suboptimal on some processes.
//   - the 6 bytes of data will be at byte offsets 10 to 15, and bytes 16 to 23 will be
//     unused. This may be surprising to some users, and suboptimal in resource usage.
// One possible tweak is to declare data_ as a uint64_t[], or to use an alignas
// declaration. As always, micro-optimizations should be informed by
// microbenchmarks, showing the benefi.

```

Reply:
Also, RE custom allocator: as discussed f2f think it should be possible to templatize InlineStorage on an allocator class.